Update WireGuard COM interface to pass params instead of config

bsclifton opened the pull request:
Fixes https://github.com/brave/brave-browser/issues/37960

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used GitHub [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
TBD - this will take a while to plan out

bsclifton commented on browser/brave_vpn/win/brave_vpn_wireguard_service/status_tray/status_tray_runner.cc:
```diff
@@ -116,8 +116,10 @@ bool StatusTrayRunner::IsVPNConnected() const {
     wireguard::EnableBraveVpnWireguardService(
-        "", base::BindOnce(&StatusTrayRunner::OnConnected,
-                           weak_factory_.GetWeakPtr()));
+        // TODO(bsclifton): fix me. Why was empty string working before?
```
Need to get official binaries and test this. I can't build locally since it requires signing to run WireGuard

bsclifton replied:
OK this does not work as-is. Will need to dig in more.

bsclifton replied:
OK found why it works. If empty is passed, the last used config will be read/used.
https://github.com/brave/brave-core/blob/2e7b0028082ed35fe0f53f89ab5dd0df244935e6/browser/brave_vpn/win/brave_vpn_wireguard_service/service/wireguard_tunnel_service.cc#L146-L148

bsclifton commented on browser/brave_vpn/win/brave_vpn_wireguard_service/service/wireguard_service_runner.cc:
```diff
@@ -188,6 +188,11 @@ HRESULT WireguardServiceRunner::InitializeComSecurity() {
   }
 
+  // Don't allow guest account to instantiate.
+  if (!dacl.AddDeniedAce(Sids::Guests(), com_rights_execute_local)) {
```
This alone should fix the complaint. This permission will deny the guest account from making the COM call.

bsclifton replied:
cc: @thypon

thypon replied:
Can we allow only the current user to make the request?

bsclifton replied:
The above is a bit tricky to read - but it's restricting the COM component to only allow calls from:
- System
- Administrators
- Person who is logged in who is NOT a guest

bsclifton commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +87,77 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+
+  GURL parsed_host_url = GURL("https://" + parsed_host);
+  if (!parsed_host_url.DomainIs("guardianapp.com") &&
+      !parsed_host_url.DomainIs("sudosecuritygroup.com")) {
```
Here's the hostname validate @diracdeltas

diracdeltas replied:
@bsclifton i think you can just check the host using [DomainIs](https://source.chromium.org/chromium/chromium/src/+/main:url/url_util.h;bpv=1;bpt=1;l=208?gsn=DomainIs&gs=KYTHE%3A%2F%2FKp4CCowBa3l0aGU6Ly9jaHJvbWl1bS5nb29nbGVzb3VyY2UuY29tL2NvZGVzZWFyY2gvY2hyb21pdW0vc3JjLy9tYWluP2xhbmc9YyUyQiUyQj9wYXRoPXVybC91cmxfdXRpbC5oI040enVpVEFHblg1Q3lJbkYzMmhfaUVqLTgwZlkzWGNZcEtFRV9BaHBsbmMKjAFreXRoZTovL2Nocm9taXVtLmdvb2dsZXNvdXJjZS5jb20vY29kZXNlYXJjaC9jaHJvbWl1bS9zcmMvL21haW4_bGFuZz1jJTJCJTJCP3BhdGg9dXJsL3VybF91dGlsLmgjV0JnU043SnVsR0d6bjZUYllEMkRraEZBbEZac1E1bjgzb0tKQjU1aFhrVQ%3D%3D) and not have to prepend the https? but fine either way

bsclifton replied:
Good call! Fixed in https://github.com/brave/brave-core/pull/24261/commits/dea24007e9c5b463ab986f663106c63975a4c76a

bsclifton commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +87,77 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+  if ((*output).length() == 0) {
+    VLOG(1) << field_name << " does not have a value";
+    return false;
+  }
```
I didn't think it was worthwhile validating the public key / private key, but we can do that if we'd like.

thypon replied:
These will end up in a file, it might be worth checking if they have the correct format at least, and they are not containing spurious characters. Null-bytes, or `\n` might format inject in the file itself.

bsclifton replied:
FWIW, `base::TrimWhitespaceASCII` is called on all the fields to strip out `\n` and other chars
https://source.chromium.org/chromium/chromium/src/+/main:base/strings/string_util.h;l=296-312;drc=51d26a40f145af6de565b22a600a39e6e39184e9

thypon replied:
looks like Trim is stripping begin and end, not the middle, which is what we want to strip.

Imagine an attacker that sends a private key with the content:

```
abcdabcdabcdabcdabcdabcdabcdabcdabcd
pre_up = "echo pwned"
```

thypon replied:
When file expanded it will include both `privatekey = abcdabcdabcdabcdabcdabcdabcdabcdabcd` and `pre_up = "echo pwned"`

bsclifton replied:
Fixed in https://github.com/brave/brave-core/pull/24261/commits/a63932075089a42f15ed93c0c66320fadcdf2175

diracdeltas reviewed (changes requested):
see comment here: https://hackerone.com/reports/2481145#activity-28112578

thypon commented on components/brave_vpn/common/wireguard/wireguard_utils_unittest.cc:
```diff
@@ -0,0 +1,66 @@
+  EXPECT_TRUE(brave_vpn::wireguard::ValidateEndpoint(
+      L"toronto-ipsec-8.guardianapp.com", &output));
+  EXPECT_TRUE(brave_vpn::wireguard::ValidateEndpoint(
+      L"france-ipsec-1.sudosecuritygroup.com", &output));
```
```
  EXPECT_FALSE(brave_vpn::wireguard::ValidateEndpoint(
      L"france-ipsec-1\n.sudosecuritygroup.com", &output));
  EXPECT_FALSE(brave_vpn::wireguard::ValidateEndpoint(
      L"france-ipsec-1 .sudosecuritygroup.com", &output));
```

If it gets format string \n properly

bsclifton replied:
Added these tests in https://github.com/brave/brave-core/pull/24261/commits/624f5db9eabda65466e921d8836df6c8091b7921

thypon commented on browser/brave_vpn/win/brave_vpn_wireguard_service/service/brave_wireguard_manager.cc:
```diff
@@ -6,19 +6,69 @@
   }
-  *last_error = !brave_vpn::wireguard::LaunchWireguardService(config);
+
+  std::string address_str;
```
Can you remember me what the `address_str` is used for?
Isn't the `endpoint` the only required parameter to start a connection?

bsclifton replied:
This is the address the VPN host assigned when getting the credentials

thypon commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +87,87 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+    VLOG(1) << "failed decoding endpoint";
+    return false;
+  }
+
```
For all the fields Endpoint//Address//Key, I'd verify if they are defined in their alphabet with a regex, before passing to any parsing method.

- For Addresses `^[A-z0-9._-:[\]]+$` (I don't think we support empty IPs here)
- For Endpoints `^[A-z0-9._-:]+$` (I don't think we support empty endpoints here)
- For Keys `^[-A-Za-z0-9+/=]+$` (I don't think we support empty keys here)

bsclifton replied:
Added this validation in https://github.com/brave/brave-core/pull/24261/commits/624f5db9eabda65466e921d8836df6c8091b7921

bridiver commented on browser/brave_vpn/win/brave_vpn_wireguard_service/service/brave_wireguard_manager.cc:
```diff
@@ -6,19 +6,69 @@
+                                     address && wcslen(address) == 0 &&
+                                     endpoint && wcslen(endpoint) == 0;
+  if (reconnect_using_last_config) {
+    *last_error = !brave_vpn::wireguard::LaunchWireguardService(L"");
```
do we expect this to happen under some circumstances or are we trying to handle an unexpected condition here? If it's the latter I think we should just fail here because we might be unintentionally opening another attack vector

bsclifton replied:
This is the current behavior - I thought I made comments in this PR explaining... but basically the tray program is passing empty string for the arguments. This triggers the "reconnect using the last known config" logic.

We could rework the tray program - but that's a bigger change of course.

thypon replied:
I think, best is to harden this part with the follow-up in https://github.com/brave/brave-browser/issues/39229

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +88,102 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+    return false;
+  }
+
+  base::TrimWhitespaceASCII(*output, base::TRIM_ALL, output);
```
so backtracking a bit on the comments that were made in a previous version about using chromium utilities, given that this is untrusted input, I think the first thing we need to do is pass it through re2 assuming that @thypon agrees that the regex results in safe output.

```
As another special case, we trust the [RE2](https://cs.chromium.org/chromium/src/third_party/re2/README.chromium) regular expression library to evaluate untrustworthy patterns over untrustworthy input strings, because its grammar is sufficiently limited and hostile input is part of the threat model against which it's been tested for years. It is not the case, however, that text matched by an RE2 regular expression is necessarily “sanitized” or “safe”. That requires additional security judgment.
```

thypon replied:
We are gonna remove `TrimWhitespaceASCII` altogether

bsclifton replied:
Removed.

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +88,102 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+  return true;
+}
+
+bool ValidateAddress(const wchar_t* address, std::string* output) {
```
use absl::optional instead of out params

bsclifton replied:
Fixed in latest commit

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils.cc:
```diff
@@ -82,6 +88,102 @@ WireguardKeyPair GenerateNewX25519Keypair() {
+    return false;
+  }
+
+  if (!url::DomainIs(parsed_host, "guardianapp.com") &&
```
we don't want to use `net::registry_controlled_domains::SameDomainOrHost` here?

bsclifton replied:
`SameDomainOrHost` requires a `GURL`. We could check that - but then I'd need to provide a schema (ex: `https://`) to the hostname for it to be valid. GURL checks fail without a schema. I had it as a GURL before but removed after @diracdeltas comment https://github.com/brave/brave-core/pull/24261#discussion_r1645201955

bridiver replied:
there is a method that takes a host, but as long as this works correctly with subdomains then it's fine

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils_unittest.cc:
```diff
@@ -0,0 +1,81 @@
+
+TEST(BraveVPNWireGuardUtilsUnitTest, ValidateAddress) {
+  std::string output;
+  // Invalid.
```
Seems like this should check things like `1111`, `1.1.1`, `1.1.1.1.1`, `300.1.1.1`?

thypon replied:
@bridiver `1.1.1` `1111` are valid IPv4, btw

bsclifton replied:
I'll add these in next commit

bsclifton replied:
and you're right @thypon - tests with those two do fail. I added the others though (1.1.1.1.1, 300.1.1.1)

bridiver replied:
@thypon they may technically be valid, but we don't expect anything like that from guardian and if we did get something like that I think we should treat it as an error

bridiver replied:
now with IPAddress::IsValid we should just follow whatever rules it enforces

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils_unittest.cc:
```diff
@@ -0,0 +1,81 @@
+      L"fe80::1ff:fe23:4567:890a", &output));
+
+  // Valid.
+  EXPECT_TRUE(brave_vpn::wireguard::ValidateAddress(L"192.168.1.1", &output));
```
do we always expect this to be in a particular IP range? Will it always be from known private IP ranges?

bsclifton replied:
No, it should be a public IP. I just put a valid IPv4 address to pass validation

bridiver commented on components/brave_vpn/common/wireguard/wireguard_utils_unittest.cc:
```diff
@@ -0,0 +1,81 @@
+  EXPECT_FALSE(brave_vpn::wireguard::ValidateEndpoint(
+      L"toronto-ipsec-8.not-guardianapp.com", &output));
+  EXPECT_FALSE(
+      brave_vpn::wireguard::ValidateEndpoint(LR"(france-ipsec-1
```
why is this one false?

bsclifton replied:
because there's a CRLF in the middle of the string

thypon replied:
It contains a carriage return, in between the string. It's invalid

bridiver commented on browser/brave_vpn/win/brave_vpn_wireguard_service/status_tray/status_tray_runner.cc:
```diff
@@ -116,8 +116,10 @@ bool StatusTrayRunner::IsVPNConnected() const {
-        "", base::BindOnce(&StatusTrayRunner::OnConnected,
-                           weak_factory_.GetWeakPtr()));
+        // passing empty params will reconnect using last known good config.
+        "", "", "", "",
```
how did this work before given that the config file was deleted?

bsclifton replied:
I was wrong about the config being deleted ☹️ I was looking in the wrong directory. Since the service runs as `LOCALSYSTEM`, it's storing at `C:\Windows\Temp\BraveVpn`

bsclifton commented on components/brave_vpn/common/wireguard/wireguard_utils_unittest.cc:
```diff
@@ -0,0 +1,81 @@
+  EXPECT_FALSE(brave_vpn::wireguard::ValidateAddress(L"a.b.c.d", &output));
+  // IPv6 not allowed.
+  EXPECT_FALSE(brave_vpn::wireguard::ValidateAddress(
+      L"fe80::1ff:fe23:4567:890a", &output));
```
IPv6 not supported is a vendor limitation. It should work just fine and we could allow it if we'd like. It's just not possible to do currently. I think this test case came up from an issue reporter

bridiver commented on browser/brave_vpn/win/wireguard_utils_win.cc:
```diff
@@ -79,14 +83,21 @@ bool EnableBraveVpnWireguardServiceImpl(const std::string& config) {
   DWORD error_code = 0;
   if (FAILED(service->EnableVpn(
-          base::UTF8ToWide(base::Base64Encode(config)).c_str(), &error_code))) {
+          ::SysAllocString(base::UTF8ToWide(server_public_key).c_str()),
```
why are we calling `SysAllocString` here when we apparently didn't have to before?

bridiver replied:
it's particularly concerning since I don't see any calls to `SysFreeString` so it seems like this will leak?

bsclifton replied:
Fixed in https://github.com/brave/brave-core/pull/24261/commits/c48a27fc395cedc277ccced2fab406a8ca5c13e8

bsclifton commented:
NOTE: as-is, this code does not work. I tried the latest PR builder and am running into problems.

I'm going to try some reverts to see if I can narrow down the issue. The last known good build was with https://github.com/brave/brave-core/commit/7dc2e0ad2dc71ca2554d98c83bc60a7063e353d9

The validation isn't too much different. I suspect the change is in `browser/brave_vpn/win/brave_vpn_wireguard_service/service/wireguard_service_runner.cc` with:
```
if (!dacl.AddDeniedAce(Sids::Guests(), com_rights_execute_local)) {
    return E_ACCESSDENIED;
}
```

bsclifton commented:
Issue solved - build works locally 🎉 

Investigating the COM interface defined here:
https://github.com/brave/brave-core/blob/9f8f2e52eb812c43b4fa0065643ef6531da65080/components/brave_vpn/common/wireguard/win/brave_wireguard_manager_idl.idl#L7

There is already a separate CLSID per-channel (Release, Beta, Nightly, etc):
https://github.com/brave/brave-core/blob/9f8f2e52eb812c43b4fa0065643ef6531da65080/browser/brave_vpn/win/service_details.cc#L29-L34

The next step I'm looking at now is making a per-channel IID (interface ID). We could then update the code to return the correct IID per channel here:
https://github.com/brave/brave-core/blob/9f8f2e52eb812c43b4fa0065643ef6531da65080/browser/brave_vpn/win/service_details.cc#L102-L104

And then the IDL can be rebuilt. Basically, if this gets installed, the IID registration (where # of arguments changes) will break other channels trying to use VPN on the machine. If this was merged and installed for Nightly, Release channel would stop working properly because the COM interface changed.

bsclifton commented:
OK made some changes and verified it works with latest PR builder. 
Rebased and squashed those changes. 
Also removed the SKIP labels so this can run on all platforms

bsclifton commented:
Tested and this works great 😄

thypon reviewed (approved):
https://bravesoftware.slack.com/archives/C079KC463FH/p1719482602829669
